src/app/sidebar/sidebar.tsx

```diff
@@ -0,0 +1,130 @@
+import classnames from "classnames";
+import React, { useState } from "react";
+import css from "./Sidebar.module.css";
+import {
+    Sidenav,
+    SidenavItems,
+    SidenavParent,
+    SidenavLink,
+    SidenavFilter,
+    OnChangeCallback,
+    OnChangeInput,
+} from "./sidenav";
+import { NavLink, useLocation, matchPath } from "react-router-dom";
+import styles from "./Sidebar.module.css";
+import i18n from "@dhis2/d2-i18n";
+import { HidePreventUnmount } from "../../components/utils";
+import { LinkItem, SidebarLinks, sidebarLinks } from "./sidebarLinks";
+
+interface SidebarNavLinkProps extends LinkItem {
+    disabled?: boolean;
+    end?: boolean;
+}
+
+const SidebarNavLink = ({ to, label, disabled, end }: SidebarNavLinkProps) => {
+    return (
+        <SidenavLink
+            to={to}
+            disabled={disabled}
+            LinkComponent={NavLink}
+            label={label}
+            end={end}
+        />
+    );
+};
+
+interface SidebarParentProps {
+    isFiltered: boolean;
+    label: string;
+    links: LinkItem[];
+}
+
+const SidebarParent = ({
+    label,
+    links = [],
+    isFiltered,
+}: SidebarParentProps) => {
+    const { pathname } = useLocation();
+
+    // Check if any of the children match the current path
+    // If they do, parent should be open by default
+    const routePathMatch = links.some((link) => {
+        return matchPath(link.to, pathname);
+    });
+
+    const forceOpen = (isFiltered && links.length > 0) || undefined;
+    return (
+        <SidenavParent
+            label={label}
+            initialOpen={routePathMatch}
+            forceOpen={forceOpen}
+        >
+            {links.map(({ to, label }) => (
+                <SidebarNavLink key={label} to={to} label={label} />
+            ))}
+        </SidenavParent>
+    );
+};
+
+export const Sidebar = ({ links = sidebarLinks }: { links: SidebarLinks }) => {
+    const [filterValue, setFilterValue] = useState("");
+
+    const handleFilterChange = (input: OnChangeInput) => {
+        setFilterValue(input.value);
+    };
+
+    const isFiltered = filterValue !== "";
+    const filteredSidebarLinks = isFiltered
+        ? Object.entries(links).reduce((acc, [key, { label, links }]) => {
+              const filteredLinkItems = links.filter(({ label }) =>
+                  label.toLowerCase().includes(filterValue.toLowerCase())
+              );
+              acc[key] = { label, links: filteredLinkItems };
+              return acc;
+          }, {} as typeof sidebarLinks)
```

Comment:
Creating the filter result is a bit hard to parse and the reduce function is doing multiple things at one. I think it's easier to grasp what's going on with `.map` and `Object.fromEntries` as that's more declarative and less imperative:

```js
Object.fromEntries(
    Object
        .entries(links)
        .map(([key, { label, links }]) => [
            key,
            {
                label,
                links: links.filter(({ label }) =>
                    label.toLowerCase().includes(filterValue.toLowerCase())
                ),
            }
        ])
)
```

Reply:
Did a refactor now, that more resembles something like this.